Grouped convolution forward device implementation and base flavors for RDNA3/4

krithalith opened the pull request:
This pull request adds support for WMMA-based convolution forward (conv fwd) by new instances with also ScaleAdd operations on A and B tensors, targeting multiple data types (fp16, bf16, int8). It introduces new example executables, updates the common header to select the correct device implementation based on a macro, and makes several improvements and fixes in the core library to support these new examples and improve robustness.

**New WMMA-based convolution examples:**

* Added three new example executables for WMMA-based conv fwd with ScaleAdd on A and B:
  - `conv_fwd_wmma_cshufflev3_scaleadd_ab_fp16.cpp`
  - `conv_fwd_wmma_cshufflev3_scaleadd_ab_bf16.cpp`
  - `conv_fwd_wmma_cshufflev3_scaleadd_ab_int8.cpp`
  [[1]](diffhunk://#diff-9b6ca395c217d0bf27e1fac0fc4865faa55fd7f54d76258ef26ee3cd0a4a03d7R1-R27) [[2]](diffhunk://#diff-2c5b2988937a763d6e212697ad1e07169d22f2e0af669c9dba385416ea000261R1-R27) [[3]](diffhunk://#diff-1bed0325d14d04f02db2b9cb0472ae13cb42403610a8fb81fef8c7d25a8c9a7bR1-R27)
* Updated `CMakeLists.txt` to build and manage dependencies for the new WMMA-based examples under a custom target.

**Device implementation selection and configuration:**

* Modified `convnd_fwd_activ_multi_ab_common.hpp` to select between WMMA and non-WMMA device implementations using the `EXAMPLE_USE_WMMA` macro, and defined a new WMMA-based `DeviceGroupedConvNDMultiABFwdInstance` with appropriate template parameters and block sizes. [[1]](diffhunk://#diff-452f8a535ad76cd62ce1a5e59178239cf59712517e57d63046f35558bf061198R12-R16) [[2]](diffhunk://#diff-452f8a535ad76cd62ce1a5e59178239cf59712517e57d63046f35558bf061198R48-R103) [[3]](diffhunk://#diff-452f8a535ad76cd62ce1a5e59178239cf59712517e57d63046f35558bf061198R157)

**Core library improvements and bug fixes:**

* Fixed uninitialized array members in `ComputePtrOffsetOfStridedBatch` by initializing arrays and scalars with `{}` in `device_grouped_conv_utils.hpp`. [[1]](diffhunk://#diff-3b727fdb13736da405ae05b3e08820f90b5eb60998297f842a44c4f8e8754b7cL201-R204) [[2]](diffhunk://#diff-3b727fdb13736da405ae05b3e08820f90b5eb60998297f842a44c4f8e8754b7cL256-R259)
* Added new overloads and improved host/device compatibility for grid descriptor and grid size calculation methods in `gridwise_gemm_wmma_cshuffle_v3_common.hpp`, and improved validity checking logic for pipeline versions. [[1]](diffhunk://#diff-94e6c276596fd990c716331b4c6be0fa93774777bd30219f37a8f5d9eda493f4R347-R360) [[2]](diffhunk://#diff-94e6c276596fd990c716331b4c6be0fa93774777bd30219f37a8f5d9eda493f4L609-R621) [[3]](diffhunk://#diff-94e6c276596fd990c716331b4c6be0fa93774777bd30219f37a8f5d9eda493f4L709-R723) [[4]](diffhunk://#diff-94e6c276596fd990c716331b4c6be0fa93774777bd30219f37a8f5d9eda493f4L724-R738) [[5]](diffhunk://#diff-94e6c276596fd990c716331b4c6be0fa93774777bd30219f37a8f5d9eda493f4R928-R929) [[6]](diffhunk://#diff-94e6c276596fd990c716331b4c6be0fa93774777bd30219f37a8f5d9eda493f4R941)
* Implemented a new `Run` method in `GridwiseGemm_wmma_cshuffle_v3` to support convolution use cases where grid descriptors are passed as arguments.

**Minor improvements:**

* Added a print statement to indicate when verification is running in the example code.

## Proposed changes

Please describe the motivation behind the pull request, whether it enables a new feature or fixes a bug. If there are associated pull requests or issues, please link them to the pull request.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

ex-rzr committed:
Fix strides of a/b_thread_desc

The descriptors are larger than needed (even though the compiler don't alloc registers for unused values).

ex-rzr committed:
Load in M/NRepeat dims with thread copy's slice instead of a loop

ex-rzr committed:
Clone BlockwiseGemmXdlops_pipeline_v1 for WMMA implementation

ex-rzr committed:
Implement Intrawave and Interwave variants of pipeline v1

ApoorvaKalyani committed:
example for gemm_universal_bf16

ApoorvaKalyani committed:
Adding examples for gemm_wmma instances

ApoorvaKalyani committed:
Added the  missing parameters

ApoorvaKalyani committed:
Fixed review comments and added executable to cmakeLists

ApoorvaKalyani committed:
Fixing clang format

ApoorvaKalyani committed:
Fixing build erros

ApoorvaKalyani committed:
Fixed compilation failure.

ApoorvaKalyani committed:
Modified some code as per gemm_universal_examples

ApoorvaKalyani committed:
Fixed the gemm specialization error

ApoorvaKalyani committed:
Fixed the build errors.

ex-rzr committed:
Add instances for Interwave and Intrawave v1

ex-rzr committed:
Add instances with ABlockLdsExtraM and BBlockLdsExtraN = 0

ex-rzr committed:
Remove instances that are too slow (mostly because of register spilling)

ancahamuraru committed:
Fixed typos for padded instances

ex-rzr committed:
Add a workaround for fp8/bf8->f32 packed conversion issue

ancahamuraru committed:
Added tests for fp16, KM_KN and KM_NK

ex-rzr committed:
Add instances for Interwave and Intrawave v1

ex-rzr committed:
Enable profiling of mixed precision with f8 and int4 on WMMA

ancahamuraru committed:
Padding not supported for when BDataType is pk_i4_t. Added fix for correct check and removed padding instances.

ancahamuraru committed:
Fixed typos

ancahamuraru committed:
Updated the set of tests for FP16

ancahamuraru committed:
Updated the set of tests for FP16

ancahamuraru committed:
Fix typo

ancahamuraru committed:
Merge branch '33-wip' of projects.streamhpc.com:amd/ai/composable_kernel into 33-wip

ancahamuraru committed:
Moved f16xi4 test under the correct data layout group

ex-rzr committed:
Fix segfault in profiler when B is pk_i4_t

b_device_buf's size in bytes is larger than b_k_n_permute so b_device_buf.ToDevice reads out-of-bounds.

ex-rzr committed:
Remove instances that are too slow (mostly because of register spilling)

ex-rzr committed:
Add missing add_device_gemm_wmma_universal_f8_f8_bf16 declarations

ex-rzr committed:
Add test case for bf16_i4

ex-rzr committed:
Add missing Regular tests

ex-rzr committed:
Use PermuteB with f16_i4 in most instances (as xdl)

Some instances use PermuteB = false for checking correctness.
See also the previous commit.

ex-rzr committed:
Fix cache flushing for pk_i4

ex-rzr committed:
Add mixed precision examples

ex-rzr committed:
Add test_gemm_universal_xdl/wmma_fp16 to REGRESSION_TESTS

They take more than 30 seconds

ex-rzr committed:
Fix a bug that fp16_i4 validation passes only with PermuteB

A permutation required by conversion from pk_i4_t to half_t does not
depend on PermuteB, they can be used independently.

ex-rzr committed:
Disable all tests and instances with f8 on gfx11

Even though f8_f16 and f16_f8 don't require f8 WMMA instructions,
gfx11 still lacks hardware instructions for fast f8->f32 conversion.

ex-rzr committed:
Add FP16 KM_NK and KM_KN test suites for XDL

These tests were added to common .inc for better testing of WMMA instances

ApoorvaKalyani committed:
adding gemm_add wmma_cshuffle and other support


(cherry picked from commit ec447e7f564095ea969eddc39ec77b843aa52976)

Co-authored-by: Cenxuan <cenxuan@streamhpc.com>

ApoorvaKalyani committed:
add instances into camkelists


(cherry picked from commit 23bf2d2771c939ea3ca7f493433c55255bffd08e)

Co-authored-by: Cenxuan <cenxuan@streamhpc.com>

ApoorvaKalyani committed:
This is work in progress, edited the template parameters in order to build

(cherry picked from commit b4fde8a3314cb44659c4bbda35f1a0133c63dc41)

Co-authored-by: Cenxuan <cenxuan@streamhpc.com>

ApoorvaKalyani committed:
temp work saved, changed the BDataType to f16 or bf16 since wmma currently not support non-equal A and B datatype


(cherry picked from commit 22fbd68f1db458ab50780a394ee2544c7a1484d1)

Co-authored-by: Cenxuan <cenxuan@streamhpc.com>

ApoorvaKalyani committed:
added datatype and use clang-format-12


(cherry picked from commit ae4e853682ef1bb27784b2f965b4a66b3751ceec)

Co-authored-by: Cenxuan <cenxuan@streamhpc.com>

ApoorvaKalyani committed:
Fixing build errors

ex-rzr committed:
Support multiple D in GridwiseGemm_wmma_cshuffle_v3

DeviceGemm_Wmma_CShuffleV3 is changed for new template parameters.

ex-rzr committed:
Support multiple D in GridwiseGemm_wmma_cshuffle_v3

DeviceGemm_Wmma_CShuffleV3 is changed for new template parameters.

ex-rzr committed:
Use ThreadGroupTensorSliceTransfer_v7r3

ex-rzr committed:
Use ThreadGroupTensorSliceTransfer_v7r3

ex-rzr committed:
Clone for device_gemm_wmma_cshuffle_v3.hpp for future Multiple D support

ex-rzr committed:
Clone for device_gemm_wmma_cshuffle_v3.hpp for future Multiple D support

ex-rzr committed:
Clone example/65_gemm_multiply_multiply/gemm_add_add_xdl_fp16.cpp for wmma

ex-rzr committed:
Clone example/65_gemm_multiply_multiply/gemm_add_add_xdl_fp16.cpp for wmma

ex-rzr committed:
Implement DeviceGemmMultipleD_Wmma_CShuffleV3

ex-rzr committed:
Implement DeviceGemmMultipleD_Wmma_CShuffleV3

ex-rzr committed:
Make gemm_add_add_wmma to work with DeviceGemmMultipleD_Wmma_CShuffleV3

ex-rzr committed:
Make gemm_add_add_wmma to work with DeviceGemmMultipleD_Wmma_CShuffleV3

Commit:
removed unnecessary ck parts from compilation

Commit:
initial gemm_add_multiply instance implementations

Commit:
fixed profiler help message for gemm_add_multiply

Commit:
improved multiply_add profiler layout help

ex-rzr committed:
Prepare gemma_add tests for adding wmma

ex-rzr committed:
Prepare gemma_add tests for adding wmma

ex-rzr committed:
Add gemm_add_fastgelu instances and test

ex-rzr committed:
Add gemm_add_fastgelu instances and test

ex-rzr committed:
Add a special wrapper to use DeviceGemmMultipleD_Wmma_CShuffleV3 with old API

ckProfiler uses DeviceGemmMultipleD (tests also call its functions), the wrapper allows to use
DeviceGemmMultipleDSplitK instances there.

ex-rzr committed:
Add a special wrapper to use DeviceGemmMultipleD_Wmma_CShuffleV3 with old API

ckProfiler uses DeviceGemmMultipleD (tests also call its functions), the wrapper allows to use
DeviceGemmMultipleDSplitK instances there.

Commit:
fixed template arguments for test instances

Commit:
added test for gemm_add_multiply

ex-rzr committed:
Merge branch '52-implement-multipled-in-gemm-universal' into 'feature/multiple-d-gemms'

DeviceGemmMultipleD_Wmma_CShuffleV3

See merge request amd/ai/composable_kernel!21

Commit:
switched to splitK interface

Commit:
log print added to splitk benchmarks

Commit:
revert main cmake comments

Commit:
Merge remote-tracking branch 'origin/feature/multiple-d-gemms' into 8-implement-device_gemm_add_multiply-for-rdna4

Commit:
newline change reverted

Commit:
added add_fastgelu instances

Commit:
revert unintended change in xdl add_fastgelu

ApoorvaKalyani committed:
Added instances for v3

ApoorvaKalyani committed:
Adding instances and executables

Commit:
created gemm_add_add_fastgelu instances

Commit:
created fastegelu instances

Commit:
added tests for all splitk fastgelus

ApoorvaKalyani committed:
Code update of template parameters modified.

ApoorvaKalyani committed:
Renamed file.

ApoorvaKalyani committed:
Added tests.

ApoorvaKalyani committed:
Added tests.

Commit:
multiply_add instances created

Commit:
updates to add_multiply splitk instances

Commit:
splitk xdl test fixes

ApoorvaKalyani committed:
resolved error tests.

ApoorvaKalyani committed:
Fixing build errors

ApoorvaKalyani committed:
Updated comments

Commit:
added wmma multiply_multiply instances

Commit:
fixed ONLY_XDL_AND_WMMA_KERNELS tag

ApoorvaKalyani committed:
Added gemm_add examples for wmma v1 and v3

ApoorvaKalyani committed:
Merge branch '61-add-examples-for-bf16-and-fp16-instances-of-gemm_add-for-v1-and-v3-for-rdna4' of projects.streamhpc.com:amd/ai/composable_kernel into 61-add-examples-for-bf16-and-fp16-instances-of-gemm_add-for-v1-and-v3-for-rdna4

Commit:
fixed / workarounded i8 instances

Commit:
Merge branch '10-implement-device_gemm_add_fastgelu-for-rdna4' into 'feature/multiple-d-gemms'

Resolve "Implement device_gemm_add_fastgelu for RDNA4"

See merge request amd/ai/composable_kernel!23

Commit:
Merge remote-tracking branch 'origin/feature/multiple-d-gemms' into 64-implement-device_gemm_multiply_multiply_instance-for-rdna4

ApoorvaKalyani committed:
Modified the v3 code to added one fp16 bxdl instance.

ApoorvaKalyani committed:
added bf16 xdl instance.

ApoorvaKalyani committed:
removed the changes as per the MR review comment.

Commit:
fp8 instances - not tested

ApoorvaKalyani committed:
Updated tests.

ApoorvaKalyani committed:
Restored the Cmake file that was reverted by mistake during rebase.

Commit:
fixed wmma_op test

ApoorvaKalyani committed:
Updated comments.

ApoorvaKalyani committed:
Updated the template parameter description

ApoorvaKalyani committed:
Added bf16 wmma instance for add_relu

Commit:
fixed rdna4 instances

Commit:
fixed back compatibility on gfx11

ApoorvaKalyani committed:
Added f16 wmma instance and corrected bf16 instance errors.

Commit:
cleanups

Commit:
fix ckProfiler

Commit:
one more cmake fix

Commit:
added fp8 instances

ApoorvaKalyani committed:
Added instances to Cmake

ApoorvaKalyani committed:
Updated tests to ad BF16 instances as per review comment

ApoorvaKalyani committed:
Added include file and cleaned up(as per review comment)

ApoorvaKalyani committed:
Updated and optimized the example code for all types.

ApoorvaKalyani committed:
Fixed clang format

Commit:
Merge branch '63-implement-device_gemm_bilinear-for-rdna4' into 'feature/multiple-d-gemms'

Resolve "Implement `device_gemm_bilinear` for RDNA4"

See merge request amd/ai/composable_kernel!26

Commit:
Resolve "Implement `device_gemm_bilinear` for RDNA4"

Commit:
Merge remote-tracking branch 'origin/feature/multiple-d-gemms' into 8-implement-device_gemm_add_multiply-for-rdna4

Commit:
Merge remote-tracking branch 'origin/feature/multiple-d-gemms' into 64-implement-device_gemm_multiply_multiply_instance-for-rdna4

Commit:
test generalization to handle FP16 shuffle better

Commit:
added missing changes

ApoorvaKalyani committed:
Modified the template parameters to make the instances work.

ApoorvaKalyani committed:
Fixed typo in profiler

ApoorvaKalyani committed:
Added v3 instances for gemm_add_relu

Commit:
addressed core review comments

ApoorvaKalyani committed:
Added test for gemm_add_relu wmma instance

ApoorvaKalyani committed:
Cleaned up the code.

ApoorvaKalyani committed:
Added examples for gemm_add_relu

ApoorvaKalyani committed:
Fixing typo to resolve build errors.

ApoorvaKalyani committed:
Fixes applied to fix  the precision loss.

Commit:
fix billinear test after merge

ApoorvaKalyani committed:
Removed the old wmma instances.

ApoorvaKalyani committed:
Added wrapper and renamed the wmma_v3 instances

ApoorvaKalyani committed:
Updated copyrights and added wrappers.

ApoorvaKalyani committed:
Fixes applied according to review comments

ApoorvaKalyani committed:
Apply 1 suggestion(s) to 1 file(s)

Co-authored-by: Robin Voetter <robin@streamhpc.com>

ApoorvaKalyani committed:
Removed the old wmma instances.

ApoorvaKalyani committed:
Updated wrapper for the v3 instances

ApoorvaKalyani committed:
removed the old wmma examples

ApoorvaKalyani committed:
Renamed the v3 instances

ApoorvaKalyani committed:
Deleted the  gtest file added by mistake.

ApoorvaKalyani committed:
Updated thge profiler with wrapper

ApoorvaKalyani committed:
Fixed test errors.

ApoorvaKalyani committed:
Fixed the review comments

ApoorvaKalyani committed:
Fixed the if condition MACROS.

ApoorvaKalyani committed:
REVERTED THE PROFILER CHANGES

ApoorvaKalyani committed:
Revert "REVERTED THE PROFILER CHANGES"

This reverts commit 21cb98546cca10c80fea142544c3f9d0937614f0.

ApoorvaKalyani committed:
Revert "Fixed test errors."

This reverts commit 13efcc6fe1488f23ee7428df20c8aad2148f9f02.

ApoorvaKalyani committed:
Revert "Updated thge profiler with wrapper"

This reverts commit 536f86661d6e119c247cf38445b9d49f8375459a.

ApoorvaKalyani committed:
Added missing wrapper instances

ApoorvaKalyani committed:
Updated copyrights.

ApoorvaKalyani committed:
Fixed typo.

ApoorvaKalyani committed:
Fixed copyrights.

ApoorvaKalyani committed:
Updated copyrights.

ApoorvaKalyani committed:
updated copyrights.

EnricoDeg committed:
multi_abd wmma support:

 - Add multiple A and B support to multiple D implementation (gridwise level)
 - Add multi_abd GEMM (device level)
 - Add instances (xdl parity)
 - Add tests (both xdl and wmma)
 - Add examples
 - Add ckProfiler support (both xdl and wmma)

Commit:
comments on the atomics workaround

Commit:
Merge branch '64-implement-device_gemm_multiply_multiply_instance-for-rdna4' into 'feature/multiple-d-gemms'

Resolve "Implement `device_gemm_multiply_multiply_instance` for RDNA4"

See merge request amd/ai/composable_kernel!27

Commit:
Merge remote-tracking branch 'origin/feature/multiple-d-gemms' into 8-implement-device_gemm_add_multiply-for-rdna4

ApoorvaKalyani committed:
Merge branch '51-create-bf16-and-f16-instances-for-gemm_add-cshuffle_v3-for-rdna4' into 'feature/multiple-d-gemms'

Resolve "Create bf16 and f16 instances for gemm_add CShuffle_v3 for RDNA4"

See merge request amd/ai/composable_kernel!17

ApoorvaKalyani committed:
Merge branch '61-add-examples-for-bf16-and-fp16-instances-of-gemm_add-for-v1-and-v3-for-rdna4' into 'feature/multiple-d-gemms'

Resolve "Add Examples for bf16 and fp16 instances of gemm_add xdl and  wmma for v1 and v3 for RDNA4"

See merge request amd/ai/composable_kernel!24

ApoorvaKalyani committed:
Merge branch '79-add-instances-and-examples-for-device_gemm_add_relu' into 'feature/multiple-d-gemms'

Resolve "Add instances and examples for device_gemm_add_relu"

See merge request amd/ai/composable_kernel!29

Commit:
Merge remote-tracking branch 'origin/feature/multiple-d-gemms' into 8-implement-device_gemm_add_multiply-for-rdna4

Commit:
fixed cmake comment

Commit:
Merge branch '8-implement-device_gemm_add_multiply-for-rdna4' into 'feature/multiple-d-gemms'

Resolve "Implement device_gemm_add_multiply for RDNA4"

See merge request amd/ai/composable_kernel!18

EnricoDeg committed:
Fix bug in device print function

EnricoDeg committed:
Fix unused template parameter

EnricoDeg committed:
Merge branch 'develop' into feature/multiple-d-gemms

EnricoDeg committed:
Fix bug from merge

krithalith committed:
Merge remote-tracking branch 'origin/develop' into 90-prepare-an-upstream-pr-for-multipled-based-gemms

krithalith committed:
clang-format-18

EnricoDeg committed:
Fix compilation error

krithalith committed:
Add support for fwd conv in gridwise implementation. Identical to run function for bwd data.

krithalith committed:
Initial device implementation for grouped conv fwd multiABD wmma cshuffleV3. Functional but needs some fixups and extra features in the future.

krithalith committed:
Make relevant profilers print the number of valid instances to aid testing.

krithalith committed:
Add instances for all vanilla 2D and 3D flavors for f16 and bf16, only one instance per instance list to save compile time for now.  Also added incomplete set of comp instances and bias_clamp for f16 2D, just to make sure the multiple-D aspects of the device implementation are working.

krithalith committed:
Reset output buffer after each run in profile_grouped_conv_fwd_impl().

krithalith committed:
Disable sharding for the new instances for now, has tendency to lead to linker errors on repeat builds.

krithalith committed:
Add CTranspose optimization for NCHW cases just like in xdl cshuffle non-v3 device implementation.

krithalith committed:
Add instances for all 8-bit 3D vanilla grouped conv fwd types, including mixed types but with the exception of deprecated f16 comp fp8. Adapt test so we can test 8-bit and mixed types.

krithalith committed:
Add int8 instances for 2D vanilla grouped conv fwd all layouts.

krithalith committed:
Implement merged groups in device impl and add instances for merged groups 3D vanilla conv fwd

krithalith committed:
Add merged groups instances for all 2D vanilla grouped conv fwd types and layouts.

krithalith committed:
Implement multi-AB support for grouped conv fwd and add example.

krithalith committed:
Add 1D instances

krithalith committed:
Add D layout tests to IsSupportedArgument()

krithalith committed:
Add comp and mem instances for all vanilla 2D grouped conv fwd types. Skipping "x2" and "part2" instance lists, can be added later without special names if necessary.

krithalith committed:
Add comp and mem instances for vanilla 3D grouped conv fwd. Skipped 2x and part2 instances, can be added later in the same instance lists.

krithalith committed:
Add some more tests for vanilla grouped conv fwd

krithalith committed:
Add 2D bias clamp instances and tests

krithalith committed:
Add 3D bias clamp instances and tests

krithalith committed:
Add 2D and 3D clamp instances and tests

krithalith committed:
Unify problem sizes across vanilla and clamp flavor tests

krithalith committed:
Clean up device implementation: remove old todos, remove unnecessary comments and print statements, tweak description, wrap all prints in env check.

krithalith committed:
Implement rotating memory and flush cache. Requires ad-hoc buffer size calculations.

krithalith committed:
Remove wmma fp8 and bf8 instances when not targetting gfx12

krithalith committed:
Add newer instances to DEVICE_INSTANCES so the main ckProfiler can build

krithalith committed:
Remove old years for newly created files.

krithalith committed:
No need to time kernels for now.

krithalith committed:
Fixup comments

krithalith committed:
Pass struct args to Gridwise Run() function by reference.

krithalith committed:
Don't use workspace memory in the case where A needs explicit transposition but B does not.

krithalith committed:
Move calculation of rotating memory buffer sizes to Argument member functions.

krithalith committed:
After the convolution to gemm transformation, the resulting 2D tensor descriptors are not necessarily RowMajor or ColumnMajor, so things should not rely on this distinction. Therefore, pass all RowMajor to the Gridwise and use a special version of CheckValidity that does not rely on 2D tensor layouts.

krithalith committed:
Unify xdl and wmma example code for grouped conv fwd scaleadd ab

krithalith committed:
Go back to passing RCR 2D tensor layouts to gridwise gemm, and use CRC for the CTranspose case. Also remove the special convolution version of checkValidity(). It seems like no matter what 2D tensor layouts you pass to the gridwise gemm, and no matter if you are using extraMN, and no matter if you are using the convolution version of checkvalidity, the results of all tests are the same.

krithalith committed:
Add wmma scaleadd ab instances to the device factory and add a completely new scaleadd_ab gtest test for wmma cshufflev3 and xdl. Currently there is no profiler for scaleadd_ab so I made my own inside the test. Furthermore for XDL only the (NDHWGC, GKZYXC, NDHWGK) layout combination existed in the instance factory so that is the only one I added for wmma cshufflev3 and the gtest test as well. Another layout is tested in example 62, for xdl and wmma cshufflev3.

krithalith committed:
Add support for V3 pipeline (tested). To be able to support num_loop < 3 we need the fixes from the batched gemm gemm MR which was already merged upstream, so just need to rebase or merge.

krithalith committed:
Merge remote-tracking branch 'origin/develop' into 65-grouped-conv-fwd-wmma

krithalith committed:
Small post-merge fixup, everything seems to work.

krithalith committed:
Do not build or run Xdl operations with Wmma backend for now. Will be reverted before upstreaming.

krithalith committed:
Extend scaleadd_ab instance lists

krithalith committed:
Extend merged groups instance lists, including adaptations of xdl "2x" instances.

krithalith committed:
Extend "comp" instance lists, including "2x" and "part2" instances. 2x instances disabled for now since they do not compile.

krithalith committed:
Extend "mem" instance lists.

krithalith committed:
Extend regular instance lists.

krithalith committed:
Fixup comments and ignored kernel arg name

krithalith committed:
Properly use the splitN offsets for D tensors in the gridwise Run() function. Was necessary to pass the bias_clamp_large_cases test.

krithalith committed:
Make sure all strides in ComputePtrOffset are at least value initialized to avoid undefined strides. Not convinced this struct is properly initialized in other code / future code.

krithalith committed:
Re-enable sharding for wmma cshufflev3 instances

krithalith committed:
Re-enable old wmma instances

krithalith committed:
Merge remote-tracking branch 'origin/develop' into 65-grouped-conv-fwd-wmma

krithalith committed:
Post merge fix to vanilla test

krithalith committed:
Re-enable Linqun's Xdl Wmma instances

krithalith committed:
Optionally allow num_k_loop <= PrefetchStages in gridwise CheckValidity. Use this for grouped conv fwd but not in general.

krithalith committed:
Remove spurious ck_tile changes that were presumably introduced somewhere in the repeated merging from develop.

krithalith committed:
Merge remote-tracking branch 'origin/develop' into 65-grouped-conv-fwd-wmma

krithalith committed:
Merge remote-tracking branch 'origin/develop' into 65-grouped-conv-fwd-wmma

krithalith committed:
Post-merge fixes. Make sure the new gridwise gemm wmma v3 common Run function can be used. Remove splitK, and forceThreadTileTransfer for now. Also add CShuffle epilogue argument.

krithalith committed:
Disable FP8 / BF8 testing on CDNA1/2, it doesn't work anymore and needs to be either fixed or removed.

illsilin commented:
Hi @krithalith, can you please resolve conflicts and sync up to develop? We need to make sure TheRock CI passes and we don't break MIOpen.

krithalith committed:
Merge remote-tracking branch 'origin/develop' into 65-grouped-conv-fwd-wmma-upstreamable

krithalith committed:
Small post-merge fixes

krithalith committed:
Fix copyright headers

krithalith committed:
Merge branch 'develop' into streamhpc/grouped-conv-fwd-wmma

bartekxk commented on include/ck/tensor_operation/gpu/grid/gridwise_gemm_wmma_cshuffle_v3.hpp:
```diff
@@ -762,6 +762,147 @@ struct GridwiseGemm_wmma_cshuffle_v3
+        const index_t n_idx = __builtin_amdgcn_readfirstlane(blockIdx.z / karg.KBatch);
+        // const index_t k_idx =
+        //     __builtin_amdgcn_readfirstlane((blockIdx.z - n_idx * karg.KBatch) * num_k_per_block);
+
```
```suggestion

```

bartekxk commented on include/ck/tensor_operation/gpu/grid/gridwise_gemm_wmma_cshuffle_v3_common.hpp:
```diff
@@ -912,14 +926,17 @@ struct GridwiseGemm_wmma_cshuffle_v3_base
             if(num_k_loop <= BlockwiseGemmPipe::PrefetchStages)
             {
-                if(ck::EnvIsEnabled(CK_ENV(CK_LOGGING)))
+                if(!(allow_short_v3_pipe && BlkGemmPipelineVer == BlockGemmPipelineVersion::v3))
```
why we need this allow_short_v3_pipe

krithalith replied:
A while ago I fixed the blockwise V3 pipeline so it could deal with main k loop sizes of 1 and 2 (previously didn't work due to improper HasMainLoop / TailNum consideration). The IsSupported check in many gridwise structs still only allows for num_k_loop >=3 though. I made it a toggle here since other device implementations that rely on this gridwise struct + v3 pipeline are technically untested with short pipes.

bartekxk replied:
So can we remove it at now?

krithalith replied:
We could remove it but then we'd need to put it back if we ever want to use the v3 pipeline properly. As is it doesn't change anything.

bartekxk commented on include/ck/tensor_operation/gpu/thread/threadwise_tensor_slice_transfer_v7r3.hpp:
```diff
@@ -287,7 +287,8 @@ struct ThreadwiseTensorSliceTransfer_v7r3
                 using elm_vector_t = typename remove_cvref_t<decltype(elm_vectors[i])>::type;
                 elm_vectors(i).template AsType<elm_vector_t>()(I0) =
-                    oob_val ? elm_vectors(i).template AsType<elm_vector_t>()[I0] : elm_vector_t{0};
+                    oob_val ? elm_vector_t{elm_vectors(i).template AsType<elm_vector_t>()[I0]}
```
why we need addtional cast here? It should work with this

krithalith replied:
I don't remember the exact reason but it was some subtle thing that made it *not* compile as it was, for certain datatypes.

bartekxk replied:
Please double check if we need this because it is better to avoid next cast

krithalith replied:
I will re-check. It might not be needed for the types we are currently upstreaming.

krithalith replied:
In theory the compiler should be able to skip the cast if the type is already correct

krithalith replied:
I can confirm that compilation still fails without this fix, at least for f8 instances (currently checking others). You said you did want f8 / bf8 instances for the convscale etc flavors that only exist for f8bf8 right? If so we need to keep the fix. I also found these old comments from our internal review process about why this fix is necessary:

**Kiefer:** I believe that when I was implementing the 8-bit float types, this line caused a compiler error because the two branches of the conditional had different types which were freely convertible to one another (non_native_vector_base<ck::f8_ocp_t, 1> and const ck::f8_ocp_t). The compiler could not determine what the result type of the conditional should be. Since the second result of the conditional was explicitly elm_vector_t, I was pretty sure that the intent was to always have the result be elm_vector_t, which is why I added the explicit conversion to the first result of the conditional, which resolved the ambiguity for the compiler.

**Enrico:** But the conversion is already done with AsType<elm_vector_t> as far as I understand

**Kiefer:** This seems to be quite subtle. What I think happens is that elm_vectors(i).template AsType<elm_vector_t>()[I0] is not an elm_vector_t but some sort of reference or proxy object to elm_vector_t. This makes the types no longer identical in the eyes of the ternary operator and both options can be converted to f8_ocp_t or non_native_vector_base<ck::f8_ocp_t, 1>, introducing the ambiguity.

bartekxk replied:
Can we add this in constexpr only for fp8? I dont like such changes because data transfer is very sensitive part of the kernel so I would like to be safe that it will not introduce some minor not needed regression

krithalith replied:
Yes good idea, will do

krithalith replied:
Done

bartekxk commented on library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward.hpp:
```diff
@@ -766,6 +776,399 @@ struct DeviceOperationInstanceFactory<ck::tensor_operation::device::DeviceGroupe
+                         is_same_v<WeiDataType, ck::bhalf_t> &&
+                         is_same_v<OutDataType, ck::bhalf_t> &&
+                         is_same_v<AComputeType, ck::bhalf_t> &&
+                         is_same_v<BComputeType, ck::bhalf_t>)
```
Here and in a lot of cpp files:
Please dont add gnhwc_*, ngchw* layouts

krithalith replied:
This will be resolved in the instance selection MR which targets this branch. After that, we will have tuned and benchmarked instances, and only have instances for the desired NHWGCxGKYXC BF16^3 FP16^3 2D3D layouts / types / dimensionality.

bartekxk reviewed (commented):
lgtm, just minor comments

krithalith committed:
Remove commented code snippet in gridwise

Co-authored-by: Bartłomiej Kocot <barkocot@amd.com>

krithalith committed:
Limit the explicit cast added in threadwise_tensor_slice_transfer_v7r3 to only be used for f8, just in case it hurts performance.

wj-laskowski committed:
Adding tuned instace list for groupoed conv fwd (#3288)

Following flavors are updated with tuned instance list:
  - grouped_conv2d_fwd
  - grouped_conv2d_fwd_bias_clamp
  - grouped_conv2d_fwd_clamp
  - grouped_conv3d_fwd
  - grouped_conv3d_fwd_bias_clamp
  - grouped_conv3d_fwd_clamp
  - grouped_conv3d_fwd_scaleadd_ab

Re-factored instance selection:
  - removed all the unnecessary instance tuples (comp/mem/16x16/generic)
  - removed all unnecessary layouts and data types

krithalith committed:
Do not use std::remove_cvref_t, does not exist in C++17, use custom one.

wj-laskowski committed:
Splitting grouped conv fwd instances (#3449)

krithalith committed:
Disable unnecessary and failing tests related to experimental CK builder

krithalith committed:
Merge remote-tracking branch 'upstreamAMD/develop' into HEAD

krithalith committed:
Disable unnecessary ck builder experimental tests fully
